File: server/src/main/java/org/dreamexposure/discal/server/api/endpoints/v2/event/CreateEventEndpoint.java
package org.dreamexposure.discal.server.api.endpoints.v2.event;

import com.google.api.client.util.DateTime;
import com.google.api.services.calendar.model.Calendar;
import com.google.api.services.calendar.model.Event;
import com.google.api.services.calendar.model.EventDateTime;

import org.dreamexposure.discal.core.calendar.CalendarAuth;
import org.dreamexposure.discal.core.crypto.KeyGenerator;
import org.dreamexposure.discal.core.database.DatabaseManager;
import org.dreamexposure.discal.core.enums.event.EventColor;
import org.dreamexposure.discal.core.logger.LogFeed;
import org.dreamexposure.discal.core.logger.object.LogObject;
import org.dreamexposure.discal.core.object.GuildSettings;
import org.dreamexposure.discal.core.object.calendar.CalendarData;
import org.dreamexposure.discal.core.object.event.EventData;
import org.dreamexposure.discal.core.object.event.Recurrence;
import org.dreamexposure.discal.core.object.web.AuthenticationState;
import org.dreamexposure.discal.core.utils.GlobalConst;
import org.dreamexposure.discal.core.utils.ImageUtils;
import org.dreamexposure.discal.core.utils.JsonUtils;
import org.dreamexposure.discal.server.utils.Authentication;
import org.json.JSONException;
import org.json.JSONObject;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.Arrays;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import discord4j.common.util.Snowflake;

@RestController
@RequestMapping("/v2/events")
public class CreateEventEndpoint {
    @PostMapping(value = "/create", produces = "application/json")
    public String createEvent(final HttpServletRequest request, final HttpServletResponse response, @RequestBody final String rBody) {
        //Authenticate...
        final AuthenticationState authState = Authentication.authenticate(request);
        if (!authState.isSuccess()) {
            response.setStatus(authState.getStatus());
            response.setContentType("application/json");
            return authState.toJson();
        } else if (authState.isReadOnly()) {
            response.setStatus(GlobalConst.STATUS_AUTHORIZATION_DENIED);
            response.setContentType("application/json");
            return JsonUtils.getJsonResponseMessage("Read-Only key not Allowed");
        }

        //Okay, now handle actual request.
        try {
            final JSONObject requestBody = new JSONObject(rBody);

            final String guildId = requestBody.getString("guild_id");
            final int calNumber = requestBody.getInt("calendar_number");


            //okay, lets actually create the event
            final GuildSettings settings = DatabaseManager.getSettings(Snowflake.of(guildId)).block();
            final CalendarData calData = DatabaseManager.getCalendar(settings.getGuildID(), calNumber).block();

            final com.google.api.services.calendar.Calendar service = CalendarAuth.getCalendarService(settings, calData).block();
            final Calendar cal = service.calendars().get(calData.getCalendarId()).execute();

            final Event event = new Event();
            event.setId(KeyGenerator.generateEventId());
            event.setVisibility("public");

            final EventDateTime start = new EventDateTime();
            start.setDateTime(new DateTime(requestBody.getLong("epoch_start")));
            event.setStart(start.setTimeZone(cal.getTimeZone()));
            final EventDateTime end = new EventDateTime();
            end.setDateTime(new DateTime(requestBody.getLong("epoch_end")));
            event.setEnd(end.setTimeZone(cal.getTimeZone()));

            if (requestBody.has("summary"))
                event.setSummary(requestBody.getString("summary"));
            if (requestBody.has("description"))
                event.setDescription(requestBody.getString("description"));
            if (requestBody.has("color"))
                event.setColorId(EventColor.fromNameOrHexOrID(requestBody.getString("color")).getId() + "");
            if (requestBody.has("location"))
                event.setLocation(requestBody.getString("location"));
            if (requestBody.has("recur") && requestBody.getBoolean("recur")) {
                final JSONObject recur = requestBody.getJSONObject("recurrence");

                final Recurrence recurrence = new Recurrence().fromJson(recur);
                final String[] rr = {recurrence.toRRule()};

                event.setRecurrence(Arrays.asList(rr));
            }

            EventData eventData = EventData.empty();
            if (requestBody.has("image")) {
                if (ImageUtils.validate(requestBody.getString("image"), settings.isPatronGuild()).block()) {
                    //Link is good...
                    eventData = EventData.fromImage(
                        Snowflake.of(guildId),
                        event.getId(),
                        event.getEnd().getDateTime().getValue(),
                        requestBody.getString("image")
                    );

                }
            }
            //Everything supported is now checked for, lets create this on google's end now.
            final Event confirmed = service.events().insert(cal.getId(), event).execute();
            if (eventData.shouldBeSaved())
                DatabaseManager.updateEventData(eventData).subscribe();

            //If we get here, nothing errored, and everything should be created correctly...
            final JSONObject responseBody = new JSONObject();
            responseBody.put("message", "Event Successfully Created");
            responseBody.put("event_id", confirmed.getId());

            response.setContentType("application/json");
            response.setStatus(GlobalConst.STATUS_SUCCESS);
            return responseBody.toString();
        } catch (final JSONException e) {
            e.printStackTrace();

            response.setContentType("application/json");
            response.setStatus(GlobalConst.STATUS_BAD_REQUEST);
            return JsonUtils.getJsonResponseMessage("Bad Request");
        } catch (final Exception e) {
            LogFeed.log(LogObject.forException("[API-v2]", "update event err", e, this.getClass()));
            response.setContentType("application/json");
            response.setStatus(GlobalConst.STATUS_INTERNAL_ERROR);
            return JsonUtils.getJsonResponseMessage("Internal Server Error");
        }
    }
}

File: server/src/main/kotlin/org/dreamexposure/discal/server/endpoints/v2/event/CreateEventEndpoint.kt
package org.dreamexposure.discal.server.endpoints.v2.event

import discord4j.common.util.Snowflake
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import org.dreamexposure.discal.core.`object`.event.Recurrence
import org.dreamexposure.discal.core.entities.Event
import org.dreamexposure.discal.core.entities.spec.create.CreateEventSpec
import org.dreamexposure.discal.core.enums.event.EventColor
import org.dreamexposure.discal.core.extensions.discord4j.getCalendar
import org.dreamexposure.discal.core.logger.LogFeed
import org.dreamexposure.discal.core.logger.`object`.LogObject
import org.dreamexposure.discal.core.utils.GlobalConst
import org.dreamexposure.discal.server.DisCalServer
import org.dreamexposure.discal.server.utils.Authentication
import org.dreamexposure.discal.server.utils.responseMessage
import org.json.JSONException
import org.json.JSONObject
import org.springframework.http.server.reactive.ServerHttpResponse
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import org.springframework.web.server.ServerWebExchange
import reactor.core.publisher.Mono
import java.time.Instant

@RestController
@RequestMapping("/v2/event")
class CreateEventEndpoint {
    @PostMapping("/create", produces = ["application/json"])
    fun create(swe: ServerWebExchange, response: ServerHttpResponse, @RequestBody rBody: String): Mono<String> {
        return Authentication.authenticate(swe).flatMap { authState ->
            if (!authState.success) {
                response.rawStatusCode = authState.status
                return@flatMap Mono.just(Json.encodeToString(authState))
            } else if (authState.readOnly) {
                response.rawStatusCode = GlobalConst.STATUS_AUTHORIZATION_DENIED
                return@flatMap responseMessage("Read-Only key not allowed")
            }

            //Handle request
            val body = JSONObject(rBody)
            val guildId = Snowflake.of(body.getString("guild_id"))
            val calendarNumber = body.getInt("calendar_number")

            return@flatMap DisCalServer.client.getGuildById(guildId).getCalendar(calendarNumber)
                    .flatMap { calendar ->
                        var spec = CreateEventSpec(
                                start = Instant.ofEpochMilli(body.getString("epoch_start").toLong()),
                                end = Instant.ofEpochMilli(body.getString("epoch_end").toLong())
                        )

                        if (body.has("name")) {
                            spec = spec.copy(name = body.getString("name"))
                        }
                        if (body.has("description")) {
                            spec = spec.copy(description = body.getString("description"))
                        }
                        if (body.has("color")) {
                            spec = spec.copy(color = EventColor.fromNameOrHexOrId(body.getString("color")))
                        }
                        if (body.has("location")) {
                            spec = spec.copy(location = body.getString("location"))
                        }
                        if (body.has("recur") && body.getBoolean("recur")) {
                            spec = spec.copy(
                                    recur = true,
                                    recurrence = Json.Default.decodeFromString(
                                            Recurrence.serializer(), body.getJSONObject("recurrence").toString())
                            )
                        }
                        if (body.has("image")) {
                            spec = spec.copy(image = body.getString("image"))
                        }

                        calendar.createEvent(spec)
                                .map(Event::toJson)
                                .map { JSONObject().put("event", it).put("message", "Success").toString() }
                                .doOnNext { response.rawStatusCode = GlobalConst.STATUS_SUCCESS }
                                .switchIfEmpty(responseMessage("Event creation failed")
                                        .doOnNext { response.rawStatusCode = GlobalConst.STATUS_INTERNAL_ERROR })
                    }.switchIfEmpty(responseMessage("Calendar not found")
                            .doOnNext { response.rawStatusCode = GlobalConst.STATUS_NOT_FOUND }
                    )
        }.onErrorResume(JSONException::class.java) {
            it.printStackTrace()

            response.rawStatusCode = GlobalConst.STATUS_BAD_REQUEST
            return@onErrorResume responseMessage("Bad Request")
        }.onErrorResume {
            LogFeed.log(LogObject.forException("[API-v2] create event err", it, this.javaClass))

            response.rawStatusCode = GlobalConst.STATUS_INTERNAL_ERROR
            return@onErrorResume responseMessage("Internal Server Error")
        }
    }
}
